### ISSUE TITLE
fix(model): restore Q2_K_XL Vulkan sequence logit parity against token execution

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12735]
# fix(model): restore Q2_K_XL Vulkan sequence logit parity against token execution

```routing
lane: model
paths: ["internal/model/vulkan_sequence_prefill_full_model_test.go", "internal/model/qwen35_hal.go", "internal/compute/vulkan_qwen35_sequence.go"]
expected_steps: 6
```

Status: implementation pending; no fix or accepted whole-model performance result is claimed.

## Work unit

leaf

## Parent context

The original control failed during qualification of #12533 with `FAK_VULKAN_GDN_PREFILL_TILED=0`. The candidate arm was skipped. The failure blocks the same-build full-model qualification required by #12098; it does not invalidate the separately passing GDN component witness or establish that the candidate caused this divergence.

Dedupe read-back on 2026-09-10 found no exact matching issue. Closed #12105 admits retained Q2_K embedding rows to sequence prefill. Closed #12695 qualifies bounded F32-row sequence parity on its own snapshot and artifact. Open #12011 concerns missing Q2 primitive pipelines and device-dispatch evidence. Open #12265 is a general first-divergence tracing project; this leaf uses only a bounded diagnostic needed to fix the witnessed sequence failure.

## Current state

One isolated physical correctness run on AMD Radeon 8060S Graphics (RADV STRIX_HALO) failed the existing full-model test with the GDN tiled candidate disabled. Both whole P32 and split 16+16 arms reported the requested and effective route `qwen35-hybrid-sequence-prefill-v1` with `fallback=false`. The reference performs one-token prefill repeatedly and computes final logits. Each arm then executes four continuation steps.

The existing gate requires cosine >= 0.99999 and relative L2 <= 0.0001 for all five complete logit vectors, plus exact generated token IDs. Observed values in this single diagnostic run were:

| Comparison to token reference | Cosine | Relative L2 |
| --- | ---: | ---: |
| Whole P32 final prefill logits | 0.999880830938147 | 0.020824018366503955 |
| Whole P32 final continuation logits | 0.9801249890222333 | 0.2018713223319414 |
| Split 16+16 final prefill logits | 0.9998738276042117 | 0.01705552031977723 |
| Split 16+16 final continuation logits | 0.9792894326228303 | 0.20881047726563515 |

All three arms generated `[248046, 198, 248045, 248068, 198]`. Matching token IDs did not satisfy the logit oracle. The test set the Q4 cooperative candidate disabled, but the optional Q4 dispatch counter was unavailable in this binary; there is no observed-zero-dispatch claim. No full-model performance distribution was accepted, and this is not a throughput measurement.

Snapshot identity for the 2026-09-10 correctness observation:

- Public source: [568c491bd2eee6c40e655a3d272fef4876a12857](https://github.com/anthony-chaudhary/fak/commit/568c491bd2eee6c40e655a3d272fef4876a12857), the clean #12533 landing based on `3858867f25b163f3b8f57c12f88f79cf02d277b3`. The measured source archive SHA-256 is `87cdd6c51e04c078dcbd8e1bc7636e5233fbbc8a4511ed2a226d8343a15bdc8b`. Runtime/shader bytes in that measured archive are identical to the clean landing; the later changes are the ticket, independent GDN component test, and build-tool shader-count verifier constant. The physical run predates the clean commit, so its source identity remains explicitly bound to the measured archive.
- Model: `Qwen3.8-27B-UD-Q2_K_XL.gguf`, SHA-256 `fd4730dd8aad070517978752b63d530aeb1740d2283cab9fa24f1e404032ddb0`, 9,828,981,664 bytes, full dense 64-layer geometry.
- Test binary SHA-256: `9b02fb05865e70ff0cb011c33178595b4784302df6ac1332c9d1a6cb5080192f`.
- Shader bundle SHA-256: `3d43b869a3e54cac2302bed0368d154dc9ebf371989d227bf542f8c1feb0d88e`.
- Operator-held raw correctness record `fullmodel-isolated-env0-control.raw.log`, SHA-256 `31e05fb97853de486d64a54f5e823042d778457616ff8dbea049092f14a1c2f8`, contains a `fak.vulkan-sequence-full-model-default/v2` continuity record with `passed=false`.

This evidence identifies a sequence-versus-token correctness gap on the observed artifact and build. It does not yet identify the first divergent layer, operation, or tensor, or prove that a particular quantized kernel is at fault.

## Problem frame

- Centrality: Core - native prefill through the ordinary model session must preserve the reference logits and continuation state before qualification can credit it.
- P1 Context: advanced - a source/artifact-bound original control reaches the native sequence route and fails a strict existing oracle.
- P2 Net value: advanced - repair the witnessed correctness blocker so the same-build full-model gate can run before any performance or default claim.
- P3 Adaptation: preserved - retain current model/session APIs, original quality thresholds, device requirements, and the candidate default-off policy.
- P4 Operations: advanced - bind a failing base and corrected physical run to exact source, artifact, binary, route, and complete logit-vector comparisons.

For Q2_K_XL native sequence prefill / Problem: whole and split prompt execution diverge from token execution beyond the existing logit gate / Better because: the first concrete state or arithmetic mismatch is corrected at the existing path / Witness: the unchanged full-model continuity oracle passes on the same artifact and device with the candidate disabled.

## Why this is next

The control must pass before it can judge a new optimization. The captured failure is sufficient to prioritize this bounded accuracy repair; it provides no percentage attribution of prefill time.

## Core through-line

`Session` prefill -> `tryQwen35SequencePrefill` -> native `Qwen35SequencePrefill` -> final logits and four continuation steps -> compare every complete vector with repeated token execution -> source-bound passing continuity receipt.

Reproduce the failure with the existing deterministic P32 prompt, locate the earliest differing value needed to explain this case, and correct the concrete sequence/state or arithmetic defect. Add only the smallest focused regression for that cause. If the first-divergence evidence places the fix outside the declared seams, update the leaf routing before editing rather than adding an unrelated trace framework.

## Gold-plating boundary

- No relaxed cosine/L2 tolerances, token-ID-only acceptance, disabled strict assertions, or silent CPU fallback.
- No general numerical tracing framework, new model loader, format redesign, serving policy, or unrelated kernel optimization.
- No GDN/Q4 candidate default promotion, performance claim, or claim that this failure establishes either candidate as the cause.
- No broad all-model qualification campaign; first repair this exact retained Q2_K_XL full-model case and its concrete regression.

## Concrete repro witness

Build the existing Vulkan native library/shader bundle using the repository build instructions, then build the model test binary from the source under examination. Run inside an independently established isolated device window; this command itself is not a hardware admission protocol.

```sh
go test -c -tags vulkan ./internal/model -o ./model.test
env FAK_QWEN35_SEQUENCE_FULL_MODEL_WITNESS=1 \
  FAK_QWEN38_GGUF=/path/to/Qwen3.8-27B-UD-Q2_K_XL.gguf \
  FAK_QWEN38_GGUF_SHA256=fd4730dd8aad070517978752b63d530aeb1740d2283cab9fa24f1e404032ddb0 \
  FAK_VULKAN_REQUIRE_DEVICE=1 FAK_VULKAN_EXPECT_DEVICE=8060S \
  FAK_QWEN35_SEQUENCE_CONTINUITY_ONLY=1 FAK_VULKAN_GDN_PREFILL_TILED=0 \
  ./model.test -test.v -test.run '^TestVulkanFullModelSequencePrefillStandalone$'
```

The test constructs 32 prompt IDs as `(53 + i*3571) % VocabSize`, compares whole P32 and split `PrefillNoLogits(16)` then `Prefill(16)` with the token reference, and checks four subsequent steps. On the witnessed snapshot, the continuity record reports `passed=false` with the values above and the test fails. Missing artifact/device is not an accepted skip. Stamp source archive, binary, and shader-bundle digests in the test's existing provenance environment fields for the reproduction and corrected run.

## Verifiable Witness

The exact existing full-model command above must fail on the witnessed base and pass after the correction without changing its oracle. An independent tester must also add a small regression tied to the identified first mismatch, preserving nonzero carried state and split continuation where they are relevant. Software fixtures alone do not satisfy the physical criterion.

## File:Line Seams

- `internal/model/vulkan_sequence_prefill_full_model_test.go:41`: `TestVulkanFullModelSequencePrefillStandalone`, the existing required-device physical witness.
- `internal/model/vulkan_sequence_prefill_full_model_test.go:92`: strict cosine/L2 quality gate; preserve its values.
- `internal/model/vulkan_sequence_prefill_full_model_test.go:132`: deterministic whole/split/token continuity arms.
- `internal/model/vulkan_sequence_prefill_full_model_test.go:322`: `standaloneCompareArms`, compares all final-prefill and continuation logit vectors.
- `internal/model/vulkan_sequence_prefill_full_model_test.go:347`: `standaloneCompareLogits`, numeric oracle.
- `internal/model/qwen35_hal.go:951` and `:1041`: normal model sequence admission and backend invocation.
- `internal/compute/vulkan_qwen35_sequence.go:469`: native sequence execution and carried-state seam.

Coordinates are from public base `3858867f25b163f3b8f57c12f88f79cf02d277b3`; re-resolve them against the final implementation before editing.

## Blast Radius & Affected Lanes

At most two implementation packages: public `internal/model` and `internal/compute`. Affected behavior is the witnessed retained Q2_K_XL native sequence path and its continuation. Private factory/serving policy, unrelated model formats, primitive Q2 pipeline admission work, and the independently qualified opt-in GDN component proceed separately.

## Quarantined Fallback Mechanism

Keep the hardware witness opt-in and required-device, record the failed control as non-qualifying, and keep the GDN candidate default off. Existing token execution remains the comparison route. Do not mark #12098's whole-model/default gate green until an actual corrected control and subsequently required comparison pass.

## Scoped Acceptance Criteria

- [ ] [SW-VERIFIED] A bounded diagnostic identifies the first concrete sequence/token mismatch for this artifact, with a targeted failing regression and no general tracing framework.
- [ ] [SW-VERIFIED] The smallest correction preserves session APIs, route truth, carried state, and the strict existing oracle; affected package checks pass.
- [ ] [HW-WITNESSED] The exact isolated original-control command passes for whole P32 and split 16+16, all five logit vectors, and exact token IDs on the identified model artifact, with the GDN candidate disabled and no fallback.
- [ ] [HW-WITNESSED] Raw fail-to-pass logs bind full source/artifact/binary/shader identities and actual device, plus route and counter availability; missing counters remain explicitly unavailable.
- [ ] [SW-VERIFIED] The resulting source and witness are linked from this tracker; performance/default promotion remains separately gated by #12098.

## Done condition

The identified Q2_K_XL original control passes the unchanged strict whole/split sequence and four-step continuation oracle on the same physical device class, and an independent targeted regression proves the corrected cause. Matching generated token IDs alone is insufficient.

## Witness

An independent tester reproduces the original failure and observes the corrected physical continuity test, reads back the durable result, and verifies the targeted fail-to-pass regression. One correctness run is not a timing distribution or a universal model/device qualification.

## Acceptance gate

Preserve cosine >= 0.99999, relative L2 <= 0.0001, exact IDs, all complete vectors, native route admission, and no fallback. Candidate default-off and #12098 remain separate gates.

## Working spine

1. Bind the exact source/artifact/device failure and retain the dedupe results.
2. Reproduce with the existing deterministic original-control witness in an isolated device window.
3. Locate the earliest concrete divergence and add its independent targeted regression.
4. Correct the scoped sequence or carried-state cause at the existing runtime path.
5. Run affected-package checks and the unchanged physical whole/split continuity oracle.
6. Land the narrow correction with this issue and source-bound fail-to-pass evidence; leave performance/default qualification to its existing gate.

## Likely files

- `internal/model/vulkan_sequence_prefill_full_model_test.go`
- `internal/model/qwen35_hal.go`
- `internal/compute/vulkan_qwen35_sequence.go`

## Lane

Public `model`, with the existing compute seam; six steps, at most two implementation packages.

## Closure binding

The resolving signed commit must reference this tracker and its actual correctness witness, bound to the clean source identity above. The witnessed accuracy failure remains unresolved until the scoped strict continuity checks pass.

**Comment by anthony-chaudhary** [comment]
Source-bound investigation, 2026-09-12: inspected clean public `f1bcc34e0a4962231d1ee5d72364c22accd2c1c7`. No source changes or fix are claimed; acceptance criteria remain unchecked.

- Independently read back the historical raw original-control log. Its SHA-256 is exactly `31e05fb97853de486d64a54f5e823042d778457616ff8dbea049092f14a1c2f8`, matching this issue. This verifies the retained failure artifact, not a fresh reproduction.
- Software witness on that public source, through WSL: `./test.ps1 ./internal/model -run '^TestQ2KEmbeddingGather(RowMatchesScalarRef|RowsPreservesFirstMiddleLastAndRepeatedOrder)$' -count=1 -v` (FAK_FAST=0). Both existing tests passed; exit 0, package execution 0.017s. This is fixture-only embedding evidence and does not establish full-model continuity.
- Independent read-only review plus coordinator source read-back identifies an early candidate seam, not the first observed numerical mismatch: eligible Q8 GDN projections use P32/P16 cooperative batch multiplication at `internal/compute/vulkan_qwen35_sequence.go:567` and `internal/compute/vulkan.go:1449`, versus fused P1 projections at `internal/compute/vulkan_qwen35_gdn.go:42`. Actual projection eligibility and divergent values for the pinned artifact must be measured before attributing the failure.
- Existing bounded diagnostics are `TestVulkanQ8CooperativeBatchMatchesDecodeRows` and `TestVulkanQwen35GDNQ8InputProjectionMatchesComposition`. Neither was run on hardware in this investigation. The latter can skip for unavailable optional fusion; a skip is not qualification. Coordinate these with #12390 rather than duplicating a tracing framework.
- Fresh physical execution is pending authoritative isolated-run admission and packet/whole-place ownership. Existing `fak-private#696` and `fak-private#870` cover these operational gaps. Access health and a service API lease do not establish standalone isolation. No device mutation, deployment, or candidate promotion occurred.

Next checkable step: obtain the authoritative isolated P32 runner packet, reproduce the unchanged full-model control, then compare actual layer-0 projection panels before GDN recurrence if it still fails. Use the smallest demonstrated mismatch to author the regression and correction. Preserve cosine >= 0.99999, relative L2 <= 0.0001, exact IDs, all five vectors, and GDN candidate disabled. Disjoint Q4 work remains independent.

**Comment by anthony-chaudhary** [comment]
Qualification readiness after independent read-only review of the opt-in #12390 panel candidate: preserve the original Q2 control by explicitly setting `FAK_VULKAN_GDN_Q8_PANEL=0` in its isolated execution packet and recording the value in provenance. The existing full-model test does not pin or report this new environment switch; an inherited value of 1 would select different projection arithmetic and would not be the original control. Keep a candidate-enabled run separate, after the original control passes.

The reviewed panel uses F32 activations with dequantized Q8 weights; the composed Q8 path quantizes activations. Component agreement with a CPU dequantized-weight dot product does not establish repeated-token logit parity. Preserve the unchanged cosine/L2/IDs/all-five-vector oracle. No fresh physical reproduction or correction is claimed; #12735 stays open.

Review identity: candidate shader SHA-256 `5e3fc33a63bdf52c10280bd07c768fc799663e838478db4b252d4fcc626846bd`, C shim SHA-256 `2e962b932e8d64a9f8d38b0b0f2e66da6692acb129ab5888743f4fb44729e29f`. Static review found no C/shader defect in descriptor order, push constants, tail/barrier handling, integer/buffer/grid bounds, or separate output ownership. These are candidate file identities, not a landed commit or hardware qualification.
